src/chain/chain.rs

```diff
@@ -0,0 +1,331 @@
+                        // Ownership transfer should be skipped. See note at
+                        // https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#synchronization-queue-transfers
+                        let (signal, wait) = new_semaphore();
+
+                        if src_layout == dst_layout {
+                            // Semaphores creates access scope for `A::all() .. A::none() - A::none() .. A::all()`
+                            // Since no layout transition required barrier can be skipped.
+                            link.release = LinkSync::Semaphore { semaphore: signal };
+                            next.acquire = LinkSync::Semaphore { semaphore: wait };
+                        } else {
+                            // Perform layout transition before signaling semaphore.
+                            link.release = LinkSync::BarrierSemaphore {
+                                semaphore: signal,
+                                access: src_access..A::none(),
+                                layout: src_layout..dst_layout,
+                                stages: src_stages..dst_stages,
+                            };
+                            next.acquire = LinkSync::Semaphore { semaphore: wait };
+                        }
+                    }
+                    _ => {
+                        let (signal, wait) = new_semaphore();
+
+                        // Different families. Content must be preserved.
+                        // Perform ownership transfer according to
+                        // https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#synchronization-queue-transfers
+                        link.release = LinkSync::Transfer {
+                            semaphore: signal,
+                            access: src_access..A::none(),
+                            layout: src_layout..dst_layout,
+                            stages: src_stages..PipelineStage::empty(),
+                            other: next.queue,
+                        };
+                        next.acquire = LinkSync::Transfer {
+                            semaphore: wait,
+                            access: A::none()..A::all(),
+                            layout: src_layout..dst_layout,
+                            stages: PipelineStage::empty()..dst_stages,
+                            other: link.queue,
+                        };
+                    }
+                }
+            } else {
+                // No other links
+                break;
+            }
+        }
+
+        if links.iter().all(Option::is_none) {
+            warn!("Empty chain {:?}", id);
+        }
+
+        Chain { links, usage }
+    }
+}
+
+impl<A, L, U, S, W> Chain<A, L, U, S, W> {
+    /// Get reference to link by index.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if no link exists at this index.
+    ///
+    pub fn link(&self, index: usize) -> &Link<A, L, S, W> {
```

Comment:
nit: could also implement `ops::Index` and such, but `index` should probably be wrapped up into something more type safe